Add module account_bank_statement_import_coda

lmignon opened the pull request:
This module is a rewrite of l10n_be_coda from odoo master using the pycoda parser.
Expected benefits:
- The pycoda parser is a better-tested external python lib.
- The module only depends of `account_bank_statement_import` (`l10_be_coda` on master branch depends of `account_voucher`, `base_iban`, `l10n_be_invoice_bba`, `account_bank_statement_import`)
- Better separation between file parsing and mapping of these data in Odoo.
- [x] This PR depends of https://github.com/OCA/bank-statement-import/pull/11

lmignon committed:
Add module account_bank_statement_import_coda
The module is a rewrite of l10n_be_coda using the pycoda parser.

lmignon committed:
[PYLINT] Fix W1401(anomalous-backslash-in-string)

lmignon commented:
The tests fail due to https://github.com/OCA/maintainer-quality-tools/pull/169

lmignon committed:
[ADD] add README.rst file

lmignon committed:
[IMP] Improve completion from coda info
Use the Structured Communication as name else use the communication if it's provided else use '/'.
If the line is a details line use the communication from the related globalisation as name.
The module now depends of l10n_be_coda since the first one is automatically installed when installing l10n_be. The menuitem 'Import Coda File' from l10n_be_coda is  however removed to avoid confusion and force the use of the current module

sbidoul commented on account_bank_statement_import_coda/__openerp__.py:
```diff
@@ -0,0 +1,41 @@
+    'license': 'AGPL-3',
+    'depends': [
+        'account_bank_statement_import',
+        'l10n_be_coda'
```
Do we really need this dependency on l10n_be_coda?

lmignon replied:
If you need coda, you probably depend from l10n_be for your business. But if you install l10n_be, l10n_be_coda will be installed as a dependency of l10n_be. By depending of l10n_be_coda, we can safely share the same model, avoid to redefine the same field in the account_bank_statement view and hide the menu item that leads to 'old style' import wizard provided by l10n_be_coda.

sbidoul replied:
Some don't use l10n_be and l10n_be_coa_multilang instead.

By same model, you mean the "CODA Notes" field? I always thought it was weird to have a specific field for that. I'd say it could be a generic field on statement or statement line that holds additional information and could be used by any parser. So not CODA specific.

lmignon replied:
OK, I'll changes my code to remove the dependency on l10n_be_coda.
You're right for the "CODA Notes" field. I can maybe propose this new field upstream in account_banck_statement_import and use this new one.

sbidoul replied:
Having the note fields in account_bank_statement_import make sense to me. If upstream accepts.

sbidoul replied:
I see there is already a note field on bank.statement.line in the account module. Do we need more?

lmignon replied:
@sbidoul according to the febelfin specification file, an information record refers to a movement record (IOW, the 'transaction_ref' of an information statement refers the 'transaction_ref' of a movement statement) . Therefore we have a way to store the communication message from information record in the field 'note' of the related bank.statement.line. I propose to ignore the information provided by the free communication records. These don't refer to movements.

sbidoul commented on account_bank_statement_import_coda/wizard/account_bank_statement_import_coda.py:
```diff
@@ -0,0 +1,218 @@
+                parsed_date = date_parser.parse(statement.creation_date)
+                year = "%s/" % parsed_date.year
+            vals.update({
+                'name': "%s%s" % (year, statement.paper_seq_number),
```
Add the journal code as prefix?

lmignon replied:
The journal info is not available at this stage of the process. Here we are on the first step of the import process.
If we need to add the journal code as prefix, we should override '_complete_stmts_vals'. If you think we need it, I can do it.

sbidoul commented on account_bank_statement_import_coda/wizard/account_bank_statement_import_coda.py:
```diff
@@ -0,0 +1,218 @@
+                info['sequence'] = sequence
+                transactions.append(info)
+        coda_notes = []
+        for line in statement.informations:
```
In this module, 'line' has many different meanings at different places. It makes the code a bit difficult to understand.

sbidoul commented on account_bank_statement_import_coda/wizard/account_bank_statement_import_coda.py:
```diff
@@ -0,0 +1,218 @@
+                'partner_name': line.counterparty_name or None,
+                'account_number': line.counterparty_number or None,
+                'note': self.get_st_line_note_msg(line),
+                'unique_import_id': line.ref + line.transaction_ref,
```
Is line.ref a good element to make a unique line id? Perhaps the year + paper_seq_number + transaction_ref makes a more unique identifier?

sbidoul commented on account_bank_statement_import_coda/wizard/account_bank_statement_import_coda.py:
```diff
@@ -0,0 +1,218 @@
+        transactions = []
+        vals = {'balance_start': balance_start,
+                'balance_end_real': balance_end_real,
+                'date': statement.creation_date,
```
Is new_balance_date a better date?

sbidoul commented on account_bank_statement_import_coda/wizard/account_bank_statement_import_coda.py:
```diff
@@ -0,0 +1,218 @@
+        try:
+            # Matches the first 24 characters of a CODA file, as defined by
+            # the febelfin specifications
+            return re.match(r'0{5}\d{9}05[ D] {7}', data_file) is not None
```
The detection feature could be part of pycoda

lmignon committed:
[IMP] remove dependency on l10n_be_coda

lmignon committed:
[IMP] The communication of information records is now stored in the note of statement line with the same transaction_ref

lmignon committed:
[IMP] use the new_balance_date as statement date

lmignon committed:
[FIX] use year 2012

lmignon committed:
[IMP] Add journal code as prefix for statement name

lmignon committed:
[IMP] rename method

lmignon committed:
[FIX] Declare external python  dependency
The declaration prevents the install of the module if the dependency is not met

sbidoul commented:
:+1:

lmignon committed:
[ADD] oca dependency on bank-statement-import

lmignon commented:
rebased

sbidoul commented:
@lmignon the bank account used in the demo data is the same as the one in l10n_be_coda. Therefore the sanitized_acc_number constraint prevents to create it.

lmignon commented:
@sbidoul green after rebased

lmignon committed:
[FIX] Changes the account number in the demo data to honour the unicity constraint with the demo data of l10n_be_coda.

lmignon commented:
@sbidoul The problem occurred when configuring the accounting data to use the Belgium Chart of account. In this case, the l10n_be_coda is automatically installed and defined the same account number as the one in account_bank_statement_import_coda. To fix the problem I've changed the one defined in  account_bank_statement_import_coda.

sbidoul commented:
The IBAN checksum is not correct but I guess it's not an issue for this purpose.

:+1:

lmignon committed:
[IMP] Define a valid IBAN account number in demo data

lmignon commented:
@sbidoul now it's valid
![image](https://cloud.githubusercontent.com/assets/544090/8326161/5a740d6e-1a5f-11e5-926d-fa60edefd79b.png)

sbidoul committed:
normalize version number

sbidoul commented:
@lmignon any idea why tests are failing here?

lmignon commented:
@sbidoul No idea. The search on the expected statement name return no result.

sbidoul committed:
[ADD] setup.py

sbidoul committed:
[IMP] coda import: do not rely on demo data

this makes the test more robust (the previous version
did not work in presence of account_constraints, and
could not be run twice in a row on the same database)

sbidoul committed:
[FIX] coda import: _complete_statement has been renamed, apparently

sbidoul committed:
[IMP] coda import: some linting

sbidoul committed:
[IMP] coda import: update README, short headers

sbidoul commented:
I fixed the test failure (it was a real bug), and brought the module up to the latest OCA standard.

I think it's now ready to merge at last.

cc/ @lmignon @adrienpeiffer

sbidoul commented:
The test failure is unrelated. It's in the company_web module and apparently resolved by https://github.com/OCA/account-financial-reporting/pull/205.

adrienpeiffer commented:
:+1:

lmignon commented:
:+1:
